### PR TITLE
[SVN] Sage multiline input

### PR DESCRIPTION
**Pull request by woutersj**
This PR updates the Sage plugin, adapting code from the Python plugin. This way multiline input is properly processed, see bug #20.

### CODE REVIEW

File: tmpy/session/tm_sage.py [hunk 1]
```diff
@@ -70,42 +66,53 @@ def compose_output(data):
...
+        flush_verbatim('%f' % data)
+    elif isinstance (data, PSOutDummy):
+        flush_ps (data.content)
+    elif isinstance (data, FileOutDummy):
+        if (data.content is None):
+            flush_verbatim ("")
+        else:
+            flush_file (data.content)
+    else:
+        print("other")
```
Comment (da-liii) [inline tmpy/session/tm_sage.py]: for debugging purpose?

Comment (woutersj) [reply tmpy/session/tm_sage.py]: Thanks for the review. Yes, I forgot to remove that one.

**Comment by da-liii** [comment]
See https://lists.gnu.org/archive/html/texmacs-dev/2021-01/msg00042.html

Changes need to be carefully tested because of the upcoming GNU TeXmacs 2.1.

**Comment by da-liii** [comment]
The SageMath downloaded from http://iso.mirrors.ustc.edu.cn/sagemath/osx/intel/sage-9.2-OSX_10.15.7-x86_64.dmg does not work by default for me.

```
(base) ➜  ~ /Applications/SageMath/src/bin/sage --help
/Applications/SageMath/src/bin/sage-env: line 137: cd: /Users/buildbot-sage/slave/binary_pkg/build/source/SageMath/jc4b6yulaujayb9sr94ia88eourzeqip0oidmas3: No such file or directory
Warning: overwriting SAGE_ROOT environment variable:
Old SAGE_ROOT=/Users/buildbot-sage/slave/binary_pkg/build/source/SageMath/jc4b6yulaujayb9sr94ia88eourzeqip0oidmas3
New SAGE_ROOT=
Error: You must set either the SAGE_LOCAL or SAGE_SCRIPTS_DIR environment variable to run this
Error setting environment variables by sourcing '/Applications/SageMath/src/bin/sage-env';
possibly contact sage-devel (see http://groups.google.com/group/sage-devel).
```

**Comment by woutersj** [comment]
Does that happen when running SageMath on its own, or from within TeXmacs?

**Comment by da-liii** [comment]
Runing SageMath on its own. Failed to launch SageMath on macOS. I try try another dmg provided by SageMath.

File: tmpy/session/tm_sage.py [hunk 2]
```diff
@@ -24,38 +24,34 @@
...
+from tmpy.postscript import ps_out, PSOutDummy, pdf_out, FileOutDummy
 from tmpy.completion import parse_complete_command, complete
+from tmpy.protocol   import *
 from sage.all        import *
 
-__version__ = '0.8.1'
-__author__ = 'Ero Carrera'
+if py_ver == 2:
+    flush_err ("Python 2 is no longer supported, please use Python 3")
+    exit (-1)
```
Comment (da-liii) [inline tmpy/session/tm_sage.py]: It hurts the old SageMath user. Imagine a SageMath and TeXmacs user who use an old Ubuntu release shipped with old SageMath which does not support Python 3.

**Comment by da-liii** [comment]
The PR looks good to me, but removing Python 2 support is not a good idea. As far as I know, SageMath just switched to Python 3. It would be better that the PR keeps old SageMath releases work. I understand that maintaining code that works for various versions for a software is really tedious. (Thanks for your fix!) If the PR only support Python 3, I think I'd rather postpone the merge of the PR than merge it in the nearing future.

**Comment by woutersj** [comment]
I didn't consider that, TBH. The code is mostly copied from the Python plugin, so I tried to run the Python plugin in Python 2 (easier than installing an old Sage version), but it doesn't work. It gives "invalid syntax" error on any input (even something like `a=1`). So it seems supporting Python 2 for either plugins will take some more adjustments.

**Comment by woutersj** [comment]
On closer inspection this error seems to be due to the way `input()` works in v3. For v2 `raw_input()` should be used. With that change the Python plugin seems to work for v2 as well (though I need to do further testing).

I'll try the same for Sage and test with an older version that uses Python2.

**Comment by da-liii** [comment]
Here is how I maintain code that works for both Python 2 and Python 3:

https://github.com/texmacs/texmacs/blob/master/plugins/tmpy/compat.py

**Comment by woutersj** [comment]
I've checked the code with [3to2](https://pypi.org/project/3to2/), this only raised that `input` needed to be changed. Further testing with Sage 8.9 (using Python 2.7.15) showed that the import of `__builtins__` also needed to be changed.
I've now verified that this runs on both Sage 8.9 and 9.2

**Comment by da-liii** [comment]
Tested (Launch the plugin and test the session in the plugin doc) using 
+ macOS 10.15 
+ SageMath 9.1
+ Python 2
+ http://iso.mirrors.ustc.edu.cn/sagemath/osx/intel/sage-9.1-py2-OSX_10.11.6-x86_64.app.dmg

**Comment by da-liii** [comment]
Tested (Launch the plugin and test the session in the plugin doc) using
+ macOS 10.15
+ SageMath 9.2
+ Python 3.8.5
+ http://iso.mirrors.ustc.edu.cn/sagemath/osx/intel/sage-9.2-OSX_10.15.7-x86_64.app.dmg

**Review by da-liii — APPROVED** [review]
Awesome!

**Pull request closed without merging**